Honor end_of_line editorconfig option when splitting strings

06needhamt opened the pull request:
Fixes #41322 

This PR ensures that the string literal splitter honours the value in the `end_of_line` editorconfig option.

I have also added a test for this but there is an issue where all the line endings on test markup revers to CRLF even when I insert a string with LF endings. I am not sure how to resolve this problem.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs:
```diff
@@ -136,16 +136,7 @@ private bool SplitString(ITextView textView, ITextBuffer subjectBuffer, Snapshot
-            }
-
-            return false;
+            return line.Snapshot.GetText().Contains("\"");
```
This allocates. Please undo

06needhamt replied:
Done

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -46,6 +44,27 @@ private abstract class StringSplitter
+
+                // Check if there is a line ending specified in the editorconfig
+                // and ensure we honour it. 
+                if (!string.IsNullOrEmpty(newLineValue) && newLineValue == "\n")
```
IsNullOrEmpty check is not needed

06needhamt replied:
Removed

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -46,6 +44,27 @@ private abstract class StringSplitter
+                    leading: default,
+                    SyntaxKind.PlusToken,
+                    SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed));
+                }
```
Can simplify all of this to just make the token but only conditionally switch the type of elastic trivia to add

06needhamt replied:
Done

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -46,6 +44,17 @@ private abstract class StringSplitter
+                PlusNewLineToken = SyntaxFactory.Token(
+                leading: default,
+                SyntaxKind.PlusToken,
+                newLineValue == "\n" ? SyntaxFactory.TriviaList(SyntaxFactory.ElasticLineFeed) : SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed));
```
indent these three lines one leve.

06needhamt replied:
Done

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -46,6 +44,17 @@ private abstract class StringSplitter
                 _indentStyle = indentStyle;
                 CancellationToken = cancellationToken;
+
+                var options = document.GetOptionsAsync().Result;
```
SplitStringLiteralCommandHandler already handles getting and using options.  so can we move this to that code and then have it pass in the appropriate newline trivia into the splitter.

06needhamt replied:
Done

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs:
```diff
@@ -37,18 +37,25 @@ public class SplitStringLiteralCommandHandlerTests
 
+            if (lineEnding == "\n" && expectedOutputMarkup != null)
+            {
+                Assert.DoesNotContain("\r\n", expectedOutputMarkup);
```
can you assert this for the inputMarkup as well?  And also the inverse (i.e. if the lineEnding is `\r\n` that we don't have naked `\n`s in the markup (can be checked with regexes).

06needhamt replied:
Done

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -45,13 +45,20 @@ private abstract class StringSplitter
+                NewLineTrivia = newLineTrivia;
                 CancellationToken = cancellationToken;
+
+                PlusNewLineToken = SyntaxFactory.Token(
```
nit: could easily be one line.

06needhamt replied:
Done

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -10,24 +10,23 @@
             protected readonly SyntaxNode Root;
             protected readonly int TabSize;
             protected readonly bool UseTabs;
+            protected readonly SyntaxTriviaList NewLineTrivia;
```
this field seems to be unnecessary (i can't see any reads of it anywhere).

06needhamt replied:
I’ve removed it.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.cs:
```diff
@@ -154,13 +155,17 @@ private bool LineContainsQuote(ITextSnapshotLine line, int caretPosition)
 
             var root = document.GetSyntaxRootSynchronously(cancellationToken);
             var sourceText = root.SyntaxTree.GetText(cancellationToken);
+            var newLineTrivia = lineEnding == "\n" ? SyntaxFactory.TriviaList(SyntaxFactory.ElasticLineFeed) : SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed);
```
```suggestion
            var newLineTrivia = SyntaxFactory.TriviaList(lineEnding == "\n" ? SyntaxFactory.ElasticLineFeed : SyntaxFactory.ElasticCarriageReturnLineFeed);
```

CyrusNajmabadi replied:
@jasonmalinowski is this the right way for features to determien appropriate line ending?  I seem to recall you had a PR where there was some core service that could be used that would figure this out based both on settings and heuristics about the file in question.  DId that go through?  Should that be used here instead?

Thanks!

06needhamt commented:
@CyrusNajmabadi I’ve addressed all of your comments, however the tests are still failing as I’m not sure how to force line endings in test code. Can you help with this?

CyrusNajmabadi commented:
Lookin!

CyrusNajmabadi commented:
FYI, this is failing in TestHonourEndOfLineOption, which i'm guessing you added?

06needhamt commented:
>FYI, this is failing in TestHonourEndOfLineOption, which i'm guessing you added?  I can try to fix.  But is it not something you can debug/validate locally?

@CyrusNajmabadi The issue is that the line endings are CRLF in verbatim strings and I need to force the line endings to be LF for the TestHonourEndOfLineOption test but I’m not sure how to do it.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

CyrusNajmabadi commented:
@06needhamt you can merge in https://github.com/CyrusNajmabadi/roslyn/tree/patch-41322 to have this working.  Thanks! :)

06needhamt commented:
@CyrusNajmabadi I've merged it in, thanks for the help!

CyrusNajmabadi commented:
yup yup!  @sharwell This is ready for IDE review.  Thanks! :)

06needhamt commented:
Ping @sharwell

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

06needhamt commented:
Ping @roslyn-ide @sharwell for review

06needhamt commented:
Ping @sharwell for review

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/SplitStringLiteral/SplitStringLiteralCommandHandlerTests.cs:
```diff
@@ -891,5 +915,29 @@ void M()
+        var v = ""Hello "" +
+            ""[||]World!"";
+    }
+}".Replace("\r\n", "\n"), lineEnding: "\n");
```
can you add cases for `$` `@` and `@$` strings?

sharwell commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -10,18 +10,15 @@
-            protected static readonly SyntaxToken PlusNewLineToken = SyntaxFactory.Token(
-                leading: default,
-                SyntaxKind.PlusToken,
-                SyntaxFactory.TriviaList(SyntaxFactory.ElasticCarriageReturnLineFeed));
```
💭 I would expect an elastic trivia to get updated to have the correct form, without needing to separately track the newline configuration inside the fix. I wonder if we can get the `CodeAction` infrastructure to do this for us.

sharwell commented:
This has come up a few times in previous features. The intent of elastic trivia is avoiding the requirement that each individual feature need to concern itself with details like this. I verified that this feature is already using elastic trivia that is expected to always apply itself with the correct final form:

https://github.com/dotnet/roslyn/blob/5dab61aa59868a5e017f13fd0c89434e97ac21c1/src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs#L26-L29

If the feature is failing to use the correct line endings, the changes need to be made where the formatter handles elastic trivia, and not inside this feature.

Note that tests covering this scenario would of course be welcome as an independent change.
